File: app/src/main/java/com/example/markkko/povezime/app/SearchPresenter.java
package com.example.markkko.povezime.app;


import com.example.markkko.povezime.core.home.search.ISearchMVP;
import com.example.markkko.povezime.core.models.Offer;
import com.example.markkko.povezime.core.models.Search;
import com.example.markkko.povezime.core.models.User;
import com.example.markkko.povezime.core.util.SchedulerProvider;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.SingleObserver;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class SearchPresenter implements ISearchMVP.Presenter {

  private SchedulerProvider schedulerProvider;

  private ISearchMVP.Interactor searchInteractor;

  private CompositeDisposable compositeDisposable;

  private ISearchMVP.View view;

  @Inject
  public SearchPresenter(SchedulerProvider schedulerProvider, ISearchMVP.Interactor searchInteractor) {
    this.schedulerProvider = schedulerProvider;
    this.searchInteractor = searchInteractor;
  }


  @Override
  public void getSearchResults(@NotNull Search data) {
    searchInteractor.getSearchResults(data)
            .observeOn(schedulerProvider.mainThread())
            .subscribe(new SingleObserver<List<Offer>>() {
              @Override
              public void onSubscribe(Disposable d) {

              }

              @Override
              public void onSuccess(List<Offer> offers) {
                getView().showResults(offers);
              }

              @Override
              public void onError(Throwable e) {
                getView().showMessage("Newtork problem");
              }
            });
  }

  @NotNull
  @Override
  public ISearchMVP.View getView() {
    return view;
  }

  @Override
  public void setView(@NotNull ISearchMVP.View view) {
    this.view = view;
  }

  @NotNull
  @Override
  public User me() {
    return searchInteractor.me();
  }

  @NotNull
  @Override
  public CompositeDisposable getDisposables() {
    if (compositeDisposable == null) {
      compositeDisposable = new CompositeDisposable();
    }
    return compositeDisposable;
  }

  @Override
  public void setDisposables(@NotNull CompositeDisposable compositeDisposable) {
    this.compositeDisposable = compositeDisposable;
  }

  @Override
  public void dispose() {
    if (compositeDisposable != null) {
      compositeDisposable.dispose();
    }
  }
}

File: app/src/main/java/com/example/markkko/povezime/core/home/search/SearchPresenter.kt
package com.example.markkko.povezime.core.home.search

import android.util.Log
import com.example.markkko.povezime.R.string.offer
import com.example.markkko.povezime.R.string.wait
import com.example.markkko.povezime.core.base.rxTransaction
import com.example.markkko.povezime.core.models.Offer
import com.example.markkko.povezime.core.models.Search
import com.example.markkko.povezime.core.models.User
import com.example.markkko.povezime.core.util.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.Thread.sleep
import java.util.logging.Handler
import javax.inject.Inject

class SearchPresenter @Inject constructor(private val schedulerProvider: SchedulerProvider,
                                          private val searchInteractor: ISearchMVP.Interactor) : ISearchMVP.Presenter {

    override var disposables: CompositeDisposable = CompositeDisposable()

    override lateinit var view: ISearchMVP.View

    override fun getSearchResults(data: Search) {
        /*(0..700).forEach {
            offerSignleRide(data)
            sleep(150)
        }*/
        // Pozivamo model koji nam vraća Observable
        searchInteractor.getSearchResults(data)
                // Govorimo Observable na kojoj niti
                // da izvrši naredbu. Mora da bude UI nit
                // jer vršimo izmene u view-u
                .observeOn(schedulerProvider.mainThread())
                // Pretplaćujemo se na Observable i
                // definišemo šta da uradimo kada podaci stignu
                .subscribe({ view.showResults(it) }) { t ->
                    view.showMessage("Network problem")
                }
    }

    private fun offerSignleRide(search: Search) {
        rxTransaction {
            searchInteractor.getSearchResults(search)
                    .subscribeOn(schedulerProvider.backgroundThread())
                    .observeOn(schedulerProvider.mainThread())
                    .subscribe({ view.showResults(it) }, { view.showMessage(it.message!!) })
        }
    }

    override fun me(): User = searchInteractor.me()
}
